### PR TITLE
Standardize error messages

### PR DESCRIPTION
**Pull request by SYangster**
- improving error messages to use the form "_ must be x but got y"
- adjusted some unit tests to match new error messages
- standardizes errors in all files except in fuel/ (covered by Kevin)

### CODE REVIEW

File: nvflare/apis/impl/controller.py [hunk 1]
```diff
@@ -103,7 +106,9 @@ def _set_stats(self, fl_ctx: FLContext):
         """
         collector = fl_ctx.get_prop(InfoCollector.CTX_KEY_STATS_COLLECTOR, None)
         if collector:
-            assert isinstance(collector, GroupInfoCollector)
+            assert isinstance(
+                collector, GroupInfoCollector
+            ), "collector must be an instance of GroupInfoCollector, but got {}".format(type(collector))
```
Comment (YuanTingHsieh) [inline nvflare/apis/impl/controller.py:111]: if we want to ensure this check,
we can use if isinstance: ... raise TypeError

File: nvflare/apis/utils/fl_context_utils.py [hunk 2]
```diff
@@ -63,7 +63,9 @@ def generate_log_message(fl_ctx: FLContext, msg: str):
 
     peer_ctx = fl_ctx.get_peer_context()
     if peer_ctx:
-        assert isinstance(peer_ctx, FLContext)
+        assert isinstance(peer_ctx, FLContext), "peer_ctx must be an instance of FLContext, but got {}".format(
+            type(peer_ctx)
+        )
```
Comment (YuanTingHsieh) [inline nvflare/apis/utils/fl_context_utils.py:68]: Same here, we need to think if we need this check or not.

If we need: suggest change to if instance raise
If we don't: remove this whole assert statement

File: nvflare/app_common/aggregators/accumulate_model_aggregator.py [hunk 3]
```diff
@@ -67,7 +67,7 @@ def accept(self, shareable: Shareable, fl_ctx: FLContext) -> bool:
         except:
             self.log_exception(fl_ctx, "shareable data is not a valid DXO")
             return False
-        assert isinstance(dxo, DXO)
+        assert isinstance(dxo, DXO), "dxo must be DXO but got {}".format(type(dxo))
```
Comment (YuanTingHsieh) [inline nvflare/app_common/aggregators/accumulate_model_aggregator.py:70]: Same here, we need to think if we need this check or not.

If we need: suggest change to if instance raise
If we don't: remove this whole assert statement

File: nvflare/app_common/aggregators/accumulate_model_aggregator.py [hunk 4]
```diff
@@ -150,7 +150,7 @@ def aggregate(self, fl_ctx: FLContext) -> Shareable:
         for v_name in vars_to_aggregate:
             n_local_iters, np_vars = [], []
             for item in self.accumulator:
-                assert isinstance(item, _AccuItem)
+                assert isinstance(item, _AccuItem), "item must be _AccuItem but got {}".format(type(item))
```
Comment (YuanTingHsieh) [inline nvflare/app_common/aggregators/accumulate_model_aggregator.py:153]: and here.

File: nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py [hunk 5]
```diff
@@ -117,7 +117,7 @@ def accept(self, shareable: Shareable, fl_ctx: FLContext) -> bool:
         except BaseException:
             self.log_exception(fl_ctx, "shareable data is not a valid DXO")
             return False
-        assert isinstance(dxo, DXO)
+        assert isinstance(dxo, DXO), "dxo must be DXO but got {}".format(type(dxo))
```
Comment (YuanTingHsieh) [inline nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py:120]: and here.

File: nvflare/app_common/executors/multi_process_executor.py [hunk 6]
```diff
@@ -320,7 +320,7 @@ def finalize(self, fl_ctx: FLContext):
 
         # wait for all relay threads to join!
         for t in self.relay_threads:
-            assert isinstance(t, threading.Thread)
+            assert isinstance(t, threading.Thread), "t must be threading.Thread but got {}".format(type(t))
```
Comment (YuanTingHsieh) [inline nvflare/app_common/executors/multi_process_executor.py:323]: and here.

File: nvflare/app_common/filters/exclude_vars.py [hunk 7]
```diff
@@ -87,7 +95,7 @@ def process(self, shareable: Shareable, fl_ctx: FLContext) -> Shareable:
             self.log_exception(fl_ctx, "shareable data is not a valid DXO")
             return shareable
 
-        assert isinstance(dxo, DXO)
+        assert isinstance(dxo, DXO), "dxo must be DXO but got {}".format(type(dxo))
```
Comment (YuanTingHsieh) [inline nvflare/app_common/filters/exclude_vars.py:98]: and here.

File: nvflare/app_common/handlers/event_recorder.py [hunk 8]
```diff
@@ -123,9 +123,9 @@ def event_error_tag(self, fl_ctx: FLContext):
 
     def validate_prop(self, prop_name: str, req: _CtxPropReq, fl_ctx: FLContext):
         stats = fl_ctx.get_prop(self._KEY_EVENT_STATS, None)
-        assert isinstance(stats, _EventStats)
+        assert isinstance(stats, _EventStats), "stats must be _EventStats but got {}".format(type(stats))
 
-        assert isinstance(req, _CtxPropReq)
+        assert isinstance(req, _CtxPropReq), "req must be _CtxPropReq but got {}".format(type(req))
```
Comment (YuanTingHsieh) [inline nvflare/app_common/handlers/event_recorder.py:128]: these are private classes, who set the stats? req?
do we really need to check the class here?

**Review by YuanTingHsieh — CHANGES_REQUESTED** [review]
Good and detailed changes.

If the original statement is `assert(data, class)`, we need to determine if this check is necessary:

- If it is necessary, suggest changing to `if not isinstance, raise TypeError`
- If it is not, suggest removing them

**Review by nvkevlu — APPROVED** [review]
All looks good to me, thanks @SYangster. @YuanTingHsieh makes a good point about the asserts. I don't know if we need to do anything in this commit as it does not look like it will harm anything, but many assert statements were only used for coding help for auto-complete in the IDE so maybe they can be removed.

**Comment by nvkevlu** [comment]
Thanks for the update. What was the decision making process behind which asserts should be removed vs the ones changed to if and raise error? Some of these are different from what I was expecting, like I imagined checking for DXO is something we do want to do, and ClientEngineInternalSpec, ServerEngineSpec, and ServerEngineInternalSpec are more internal and ones we could remove.

Otherwise, I think the rest look good to me.

**Comment by SYangster** [comment]
For the DXOs, I saw that from_shareable() in the try-except already does some error checking and will return a DXO. I also removed some where the variables were defined in the same file rather than passed in. For a lot of the other ones I decided to err on the side of caution/ leave them to keep the code more readable with the type checking. I can remove ClientEngineInternalSpec, ServerEngineSpec, and ServerEngineInternalSpec since those are more internal though.

**Comment by yanchengnv** [comment]
Good question, Kevin.

In general, if you know that an assertion can never be wrong (e.g. the object was already created to be type X), then the assert's only use was for code autocompletion. I'm not sure we have to remove it. 

Otherwise, if the object is passed to you through a method arg, then the assert could go wrong and should checked.

**Review by nvkevlu — APPROVED** [review]
Thank you, this looks good and your comment helps explain and it makes sense. If you are removing some asserts for the specs already, we could, or we could leave them.

**Comment by SYangster** [comment]
Ok great, I think I will leave them in for now

File: nvflare/app_common/workflows/cross_site_model_eval.py [hunk 9]
```diff
@@ -125,8 +125,12 @@ def start_controller(self, fl_ctx: FLContext):
...
             self._participating_clients = [c.name for c in clients]
         else:
-            assert isinstance(self._participating_clients, list), "participating_clients must be a list"
-            assert all(isinstance(x, str) for x in self._participating_clients), "participating_clients must be strings"
+            if not isinstance(self._participating_clients, list):
+                raise TypeError(
+                    "participating_clients must be a list but got {}".format(type(self._participating_clients))
+                )
+            if not all(isinstance(x, str) for x in self._participating_clients):
+                raise TypeError("participating_clients must be strings")
```
Comment (YuanTingHsieh) [inline nvflare/app_common/workflows/cross_site_model_eval.py:133]: Is it more proper to move this part in __init__?

File: nvflare/private/event.py [hunk 10]
```diff
@@ -48,7 +48,8 @@ def fire_event(event: str, handlers: list, ctx: FLContext):
 
     if handlers:
         for h in handlers:
-            assert isinstance(h, FLComponent)
+            if not isinstance(h, FLComponent):
+                raise TypeError("h must be FLComponent but got {}".format(type(h)))
```
Comment (YuanTingHsieh) [inline nvflare/private/event.py:52]: It might be hard for users to understand what is "h"

File: nvflare/private/fed/server/shell_cmd.py [hunk 11]
```diff
@@ -152,7 +155,8 @@ def validate_shell_command(self, args: List[str], parse_result):
                 return "only one file is allowed"
 
             for f in parse_result.files:
-                assert isinstance(f, str)
+                if not isinstance(f, str):
+                    raise TypeError("f must be str but got {}".format(type(f)))
```
Comment (YuanTingHsieh) [inline nvflare/private/fed/server/shell_cmd.py:159]: also this one, but we can improve these later.

File: nvflare/private/fed_json_config.py [hunk 12]
```diff
@@ -163,7 +163,8 @@ def finalize_config(self, config_ctx: ConfigContext):
 
         data_filter_table = {}
         for c in self.task_data_filter_chains:
-            assert isinstance(c, FilterChain)
+            if not isinstance(c, FilterChain):
+                raise TypeError("c must be FilterChain but got {}".format(type(c)))
```
Comment (YuanTingHsieh) [inline nvflare/private/fed_json_config.py:167]: this one as well.

**Review by YuanTingHsieh — APPROVED** [review]

File: nvflare/private/event.py (hunk 10 above)
Comment (SYangster) [reply nvflare/private/event.py]: I agree, fixed

File: nvflare/app_common/workflows/cross_site_model_eval.py (hunk 9 above)
Comment (SYangster) [reply nvflare/app_common/workflows/cross_site_model_eval.py]: done

**Review by YuanTingHsieh — APPROVED** [review]

**Review by nvkevlu — APPROVED** [review]

**Pull request merged**